File: Data Structures & Algorithms/partition-equal-subset-sum/submission-3.java
class Solution {
    public boolean canPartition(int[] nums) {
        int n = nums.length;
        int sum = Arrays.stream(nums).sum();

        if(sum % 2 == 1){
            return false;
        }

        int target = sum / 2;

        return dfs(nums, 0, target);
    }

    private boolean dfs(int[] nums, int i, int target){
        if(target == 0){
            return true;
        }

        if(i == nums.length){
            return false;
        }

        return dfs(nums, i+1, target) || dfs(nums, i+1, target - nums[i]);
    }
}

File: Data Structures & Algorithms/target-sum/submission-3.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        dp={}

        def dfs(i, target, sum):
            if i == len(nums):
                if target == sum:
                    return 1
                return 0

            if (i, sum) in dp:
                return dp[(i, sum)]

            dp[(i, sum)] = dfs(i+1, target, sum + nums[i]) + dfs(i+1, target, sum - nums[i])
            return dp[(i, sum)]

        return dfs(0, target, 0)